### PR TITLE
Fix #199 path recursion error on windows

### LINKED ISSUE #199: ERROR    maximum recursion depth exceeded while calling a Python object
It runs into the following ERROR:

(base) c:\tmp>docker-squash -f b0aa1bedc1a4 -t test/spark:squashed test/spark:BAK
2020-09-04 07:19:07,396 root         INFO     docker-squash version 1.0.8, Docker 48a66213fe, API 1.40...
2020-09-04 07:19:07,485 root         INFO     Using v2 image format
2020-09-04 07:19:08,366 root         INFO     Old image has 36 layers
2020-09-04 07:19:08,733 root         INFO     Checking if squashing is necessary...
2020-09-04 07:19:08,733 root         INFO     Attempting to squash last 11 layers...
2020-09-04 07:19:08,735 root         INFO     Saving image sha256:7953fb24f7252c9104285fbfff6096cbe9b8e4969b36abc2f4de138387b5c42e to C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old directory...
2020-09-04 10:25:02,559 root         INFO     Image saved!
2020-09-04 10:25:05,637 root         INFO     Squashing image 'test/spark:BAK'...
2020-09-04 10:25:09,118 root         INFO     Starting squashing...
2020-09-04 10:28:34,738 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\03c3a272ee5276d21d3feb9ef6a3894483c5c9f92fe77ea1ce1760c3964fbf53\layer.tar'...
2020-09-04 10:31:26,585 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\a7b00a67fdbcc869cb929bece7823c257a6536cb976ec503e4a12c6d602a231c\layer.tar'...
2020-09-04 10:35:13,767 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\616dab28817f62361e167683fb389e68e0f91350b0b530c5d16214fb19bdde84\layer.tar'...
2020-09-04 10:35:24,827 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\cfde5ee5f8d0f174820c8cd5b11b4ee356ab09a9a5fbbdfb28c1c669f7ef5b56\layer.tar'...
2020-09-04 10:35:26,586 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\d1932b08a127c9dcd971d7ebddf7f703fc6d2032f5e3c84a1fceb293f1a10920\layer.tar'...
2020-09-04 10:35:27,368 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\dfe26d663a0845648658dae2d7b6b5deba518e7086abf6ab107b9fcfb53ba7ea\layer.tar'...
2020-09-04 10:35:27,501 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\860eff90c23384005c94fffa11c75acb2956ebb1dd415399c88ed73626d6e59d\layer.tar'...
2020-09-04 10:35:29,193 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\423bf5e4a1a89a431b3908b95062e7ab199c7156922160076195ec2f30cc3eef\layer.tar'...
2020-09-04 10:40:35,061 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\582f4a64c782c8a4fc6bcb04b2cb67bf7157a3a58d4fdce4bed87c5ec9f5e46f\layer.tar'...
2020-09-04 10:40:42,994 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\2aee4d18f193ff8c08a26dff9d4a0682d8d1380c0c1e627946b6ad8ede88cf80\layer.tar'...
2020-09-04 10:40:43,835 root         INFO     Squashing file 'C:\Users\testuser\AppData\Local\Temp\docker-squash-g_dqw_vb\old\443496af7c489aad0f7271ba5c93600df752a2d96cc33f48f4565e999a5a6005\layer.tar'...
2020-09-04 10:49:13,507 root         ERROR    maximum recursion depth exceeded while calling a Python object
2020-09-04 10:49:13,524 root         ERROR    Execution failed, consult logs above. If you think this is our fault, please file an issue: https://github.com/goldmann/docker-squash/issues, thanks!

(base) c:\tmp>python --version
Python 3.7.6

### PR DESCRIPTION
**Pull request by chaos95**
Fixes #199

- Modify _path_hierarchy to use pathlib and eliminate the recursion altogether
- Change _path_hierarchy to return the hierarchy in the order required by its calling code, without that code needing to `reversed()` the output
- Update existing unit tests to match the above
- Also modify the unit tests to be more explicit about what kind of path is being passed in
- Add a new Windows-path-specific unit test to verify that it works on Windows and fixes the issue raised in the ticket

### CODE REVIEW

**Comment by chaos95** [comment]
Not sure if I'll be able to make this work on 2.7; it relies pretty heavily on pathlib, which is a py3 package. 

Unfortunately I'm not really in a position to dedicate any time to porting it to 2.7-friendly code, so I'll understand if you'd prefer to can the PR 😬

File: docker_squash/image.py [hunk 1]
```diff
@@ -932,19 +935,15 @@ def _path_hierarchy(self, path):
```
Comment (rnc) [inline docker_squash/image.py:935]: Isn't the above comment description now wrong?

**Comment by rnc** [comment]
> Not sure if I'll be able to make this work on 2.7; it relies pretty heavily on pathlib, which is a py3 package.
> 
> Unfortunately I'm not really in a position to dedicate any time to porting it to 2.7-friendly code, so I'll understand if you'd prefer to can the PR grimacing

I've proposed dropping Python 2.7 in #219.

File: docker_squash/image.py (hunk 1 above)
Comment (chaos95) [reply docker_squash/image.py:935]: Good call. I have modified the comment to better reflect the way the function now operates.

**Review by rnc — APPROVED** [review]

**Review by goldmann — APPROVED** [review]

**Comment by philipparndt** [comment]
@goldmann can you please release a new version with this fix?

**Pull request merged**